SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverAuthorityDebugger.cpp

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/SpatialNetDriverAuthorityDebugger.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "EngineClasses/SpatialShadowActor.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialNetDriverAuthorityDebugger);
+
+const TArray<FName> USpatialNetDriverAuthorityDebugger::SuppressedActors = { TEXT("GameStateBase"), TEXT("DefaultPawn"),
+																			 TEXT("PlayerState") };
+
+const TArray<FName> USpatialNetDriverAuthorityDebugger::SuppressedProperties = { TEXT("Role"), TEXT("RemoteRole"), TEXT("Owner"),
+																				 TEXT("OwnerPC"), TEXT("CurrentStepIndex") };
+
+void USpatialNetDriverAuthorityDebugger::Init(USpatialNetDriver& InNetDriver)
+{
+	NetDriver = &InNetDriver;
+}
+
+void USpatialNetDriverAuthorityDebugger::CheckUnauthorisedDataChanges(const AActor& InActor)
+{
+	if (!NetDriver->IsServer())
+	{
+		return;
+	}
+
+	Worker_EntityId EntityId = NetDriver->PackageMap->GetEntityIdFromObject(&InActor);
+
+	USpatialShadowActor** SpatialShadowActor = SpatialShadowActors.Find(EntityId);
+
+	if (SpatialShadowActor == nullptr)
+	{
+		return;
+	}
+
+	(*SpatialShadowActor)->CheckUnauthorisedDataChanges(InActor);
+}
+
+void USpatialNetDriverAuthorityDebugger::AddSpatialShadowActor(const Worker_EntityId_Key EntityId)
+{
+	if (!NetDriver->IsServer())
+	{
+		return;
+	}
+
+	AActor* Actor = Cast<AActor>(NetDriver->PackageMap->GetObjectFromEntityId(EntityId));
+	if (!IsValid(Actor) || Actor->IsPendingKillOrUnreachable())
+	{
+		return;
+	}
+
+	if (SpatialShadowActors.Contains(EntityId))
+	{
+		UE_LOG(LogSpatialNetDriverAuthorityDebugger, Error,
+			   TEXT("Should only be adding a SpatialShadowActor once for each entity, EntityID %i"), EntityId);
+	}
+	else
+	{
+		USpatialShadowActor* SpatialShadowActor(NewObject<USpatialShadowActor>());
+		SpatialShadowActor->Init(*Actor);
+		SpatialShadowActors.Emplace(EntityId, SpatialShadowActor);
+	}
+}
+
+void USpatialNetDriverAuthorityDebugger::RemoveSpatialShadowActor(const Worker_EntityId_Key EntityId)
+{
+	if (!NetDriver->IsServer())
+	{
+		return;
+	}
+
+	SpatialShadowActors.Remove(EntityId);
+}
+
+void USpatialNetDriverAuthorityDebugger::UpdateSpatialShadowActor(const Worker_EntityId_Key EntityId)
+{
+	if (!NetDriver->IsServer())
+	{
+		return;
+	}
+
+	USpatialShadowActor** SpatialShadowActor = SpatialShadowActors.Find(EntityId);
+
+	if (SpatialShadowActor == nullptr)
+	{
+		// We can receive updates without receiving adds - in this case create the SpatialShadowActor
+		AddSpatialShadowActor(EntityId);
+		return;
+	}
+
+	AActor* Actor = Cast<AActor>(NetDriver->PackageMap->GetObjectFromEntityId(EntityId));
+	if (!IsValid(Actor) || Actor->IsPendingKillOrUnreachable())
+	{
+		RemoveSpatialShadowActor(EntityId);
```

Comment:
If we check this higher up, then we won't need to Add and then Remove?

Reply:
Redundant now due to other refactoring